virtual_rainforest/main.py

```diff
@@ -181,6 +138,80 @@ def extract_timing_details(
     return start_time, update_interval, raw_interval, end_time
 
 
+def _get_model_sequence(
+    config: Config, models: dict[str, Any], method: str
+) -> dict[str, Any]:  # FIXME dict[str, Any]-> dict[str, Type[BaseModel]]
+    """Get a tuple of the model execution sequence for a given model stage.
+
+    This function uses the ``priority`` sections in model configurations to establish an
+    execution sequence for a given model method (currently, one of ``init`` or
+    ``update``). For example, the configuration:
+
+    .. code-block:: toml
+        [plants.priorities]
+        init = ['abiotic']
+        update = ['abiotic']
+
+    would set that the ``plants`` model should be both initialised and updated before
+    the ``abiotic`` model.
+
+    Args:
+        config: A validated configuration object.
+        models: A dictionary of model subclasses or instances.
+        method: The :class:`~virtual_rainforest.core.base_model.BaseModel` method to use
+            to define the model execution sequence.
+
+    Returns:
+        The function returns a dictionary, keyed by model name, of model classes or
+        instances in the requested execution order.
+
+    Raises:
+        ConfigurationError: if the configuration priority details include a model name
+            that is not included in the configuration or if the configuration priorities
+            are cyclic.
+    """
+
+    # Extract priority information for the models for the given step, checking that the
+    # entries are sane
+    priorities: dict[str, list[str]] = {}
+    for model_name in models:
+        config_priorities = config[model_name]["priority"][method]
+
+        # Check the model doesn't have priority over itself
+        if model_name in config_priorities:
+            to_raise = f"Model {method} priorities for {model_name} includes itself"
+            LOGGER.critical(to_raise)
+            raise ConfigurationError(to_raise)
+
+        # Check the priorities are all models included in the configuration.
+        unknown_priorities = set(config_priorities).difference(models.keys())
+        if unknown_priorities:
+            to_raise = (
+                f"Model {method} priorities for {model_name} includes "
+                f"unconfigured models: {','.join(unknown_priorities)}"
+            )
+            LOGGER.critical(to_raise)
+            raise ConfigurationError(to_raise)
```

Comment:
I don't understand why you need this. As long as the `priorities` dictionary includes all the models, the sorter will work, regardless of each of the models having a complete list of priorities or not. So, the following works totally fine:

```python
priorities = {"A": ["B", "C"], "B": [], "C": [], "D": ["A"]}
sorter = TopologicalSorter(priorities )
list(reversed(list(sorter.static_order())))
```
resulting in:
```output
['D', 'A', 'C', 'B']
```
`C` and `D` have no priorities, so they are just allocated at some point when it is suitable to fulfil other model's priorities. Am I missing anything?

Reply:
This catches the case where - for example - `soil` has the priorities `['abiotic_simple', 'logging']` but  the `logging` model is not in the model configuration.

That's my last point in the PR description - I now think this step a mistake actually - we'd want a default priority list for models that includes all known priorities and these just get removed if a given model is not being used. That way we don't have to continually mess around with configuration - `soil` gets default priority over `logging` when `logging` is included, but otherwise the priorities get reduced to the configured models. @jacobcook1995? 

I wasn't expecting the code below to work, but we certainly don't want it happening:

```python
In [131]: priorities = {"A": ["B", "C", "not_a_configured_model"], "B": [], "C": [], "D": ["A"]}
     ...: sorter = TopologicalSorter(priorities )
     ...: list(reversed(list(sorter.static_order())))
Out[131]: ['D', 'A', 'not_a_configured_model', 'C', 'B']
```

Reply:
Yeah I was going to suggest adding default priorities (or dependancies) to the `JSON` schema to minimise the amount of messing around with config (which is probably unavoidable in some cases, but shouldn't be necessary to run our core set of models). I do think having a default priority list that can be overridden using the configuration is probably the better way to go though